connect/mirror/src/test/java/org/apache/kafka/connect/mirror/OffsetSyncStoreTest.java

```diff
@@ -155,6 +154,58 @@ public void testPastOffsetTranslation() {
         }
     }
 
+    // this test has been wriiten knowing the exact offsets syncs stored
+    @Test
+    public void testPastOffsetTranslationWithoutInitializationReadToEnd() {
+        final int maxOffsetLag = 10;
+
+        FakeOffsetSyncStore store = new FakeOffsetSyncStore() {
+            @Override
+            void backingStoreStart() {
+                for (int offset = 0; offset <= 1000; offset += maxOffsetLag) {
+                    sync(tp, offset, offset);
+                    assertSparseSyncInvariant(this, tp);
+                }
+            }
+        };
+
+        store.start(false);
+
+        // After starting but before seeing new offsets
+        assertTranslationsNearby(store, 400, 480, 0);
+        assertTranslationsNearby(store, 500, 720, 480);
+        assertTranslationsNearby(store, 1000, 1000, 990);
+
+        for (int offset = 1000; offset <= 10000; offset += maxOffsetLag) {
+            store.sync(tp, offset, offset);
+            assertSparseSyncInvariant(store, tp);
+        }
+
+        // After seeing new offsets, 1000 was kicked out of the store, so
+        // 1000 can only be traslated to 1, only previously stored offset is 0
+        assertTranslationsNearby(store, 1000, 3840, 0);
+
+        // We can translate offsets between the latest startup offset and the latest offset with variable precision
+        // Older offsets are less precise and translation ends up farther apart
+        assertTranslationsNearby(store, 3840, 3840, 0);
+        assertTranslationsNearby(store, 7680, 7680, 3840);
+        assertTranslationsNearby(store, 8640, 8640, 7680);
+        assertTranslationsNearby(store, 9120, 9120, 8640);
+        assertTranslationsNearby(store, 9600, 9600, 9120);
+        assertTranslationsNearby(store, 9840, 9840, 9600);
+        assertTranslationsNearby(store, 9900, 9900, 9840);
+        assertTranslationsNearby(store, 9960, 9960, 9900);
+        assertTranslationsNearby(store, 9990, 9990, 9960);
+        assertTranslationsNearby(store, 10000, 10000, 9990);
```

Comment:
These are all expressible with assertSparseSync, there's no need for a new assert method.

The offsets between the syncs (like 400, 500, and 1000) are implicitly part of the assertSparseSync calls.  When testing `assertSparseSync(store, 480, 0)`, it asserts that 479 translates to 1, which also implies that 400 would translate to 1 as well.

```suggestion
        // After starting but before seeing new offsets
        assertSparseSync(store, 480, 0);
        assertSparseSync(store, 720, 480);
        assertSparseSync(store, 1000, 990);

        for (int offset = 1000; offset <= 10000; offset += maxOffsetLag) {
            store.sync(tp, offset, offset);
            assertSparseSyncInvariant(store, tp);
        }

        // After seeing new offsets, 1000 was kicked out of the store, so
        // offsets before 3840 can only be translated to 1, only previously stored offset is 0
        assertSparseSync(store, 3840, 0);
        assertSparseSync(store, 7680, 3840);
        assertSparseSync(store, 8640, 7680);
        assertSparseSync(store, 9120, 8640);
        assertSparseSync(store, 9600, 9120);
        assertSparseSync(store, 9840, 9600);
        assertSparseSync(store, 9900, 9840);
        assertSparseSync(store, 9960, 9900);
        assertSparseSync(store, 9990, 9960);
        assertSparseSync(store, 10000, 9990);

```

Reply:
brilliant!